File: 프로그래머스/2/49993. 스킬트리/스킬트리.py
def solution(skill, skill_trees):
    
    orders = dict()
    
    for i, s in enumerate(skill):
        orders[s] = i
    
    link = dict()
    
    for i in range(len(skill)):
        link[skill[i]] = list(skill[i+1:])

    result = 0
    for tree in skill_trees:
        t = dict(orders)
        if can(t, tree, link):
            result += 1
    return result    
                    
def can(t, tree, link):
    for a in tree:
        if a in t:
            if t[a] >= 1:
                return False

            for l in link[a]:
                t[l] -= 1
    return True

File: 프로그래머스/2/49993. 스킬트리/스킬트리.java
import java.util.*;
import java.util.stream.*;


class Solution {
    public int solution(String skill, String[] skill_trees) {
        Set<Character> s = new HashSet<>();
        
        for (char c : skill.toCharArray()) {
            s.add(c);    
        }
        
        int answer = 0;
        for (String tree : skill_trees) {
            answer += isRight(s, skill, tree);
        }
        
        return answer;       
    }
    
    public int isRight(Set<Character> s, String skill, String skill_tree) {

        int index = 0;
        for (char t : skill_tree.toCharArray()) {
            if (s.contains(t)) {
                if (skill.charAt(index) != t) {
                    return 0;                  
                }
                index++;
            }
        }
        return 1;
    }
}